content/blog/2019-12-13-futil-fsm/index.md

```diff
@@ -0,0 +1,315 @@
+
+<img src="fsm.png" style="width: 100%">
+
+
+
+
+
+## Implementation
+
+<img src="flow.png" style="width: 100%">
+
+At implementation, to realize what we describe in design overview, we gradually added intermediate passes.
+
+### First Intermediate Pass
+
+There is a *Visitor* trait in this pass, which performs a recursive walk of abstract syntax tree (AST), so each individual pass can perform modification to the AST with function calls including `add_structure`, `add_input_port`, `remove_structure`, and etc provided in struct `Changes`.
+
+#### Read Wires
+
+The first pass is adding read wires. We go though all input and output ports of each component, adding corresponding *read* ports, and then each wire of the components and adding *read* wires to the ports. We do this pass ahead of creating FSM signatures because we don't want to create *read* wires for control signals like *valid* and *ready*.
+
+#### FSM Signatures
+
+This pass translates control syntax to FSM components.
+
+Based on the design logic of FSMs, we can specify the inputs and outputs of each FSM component and the wires connecting each ports to its subcomponents. Notice we also need to add *cond_read* signals to specify whether the *condition* signal from the comparison component is readable.
+
+| FSM                  | Input Ports                                   | Output Ports                 |
+| -------------------- | --------------------------------------------- | ---------------------------- |
+| `enable`/`par`/`seq` | *val, rdy_A, rdy_B, ...*                      | *rdy, val_A, val_B, ...*     |
+| `if`                 | *val, rdy_con, cond, cond_read, rdy_T, rdy_F* | *rdy, val_con, val_T, val_F* |
+| `while`              | *val, rdy_con, cond, cond_read, rdy_body*     | *rdy, val_con, val_body*     |
+
+
+
+### Interfacing
+
+This pass creates input port *clock* for all components and *valid* for the top level component. Notice in FuTIL we actually does not have the notion of logical time step. However, to make things easier for RTL translation, we created these ports.
```

Comment:
- does -> do (to agree with the subject "we")
- I'm not entirely sure what you mean by saying that we *don't* have a notion of a logical time step. Isn't a logical time step what happens in a `seq` expression in the control part?